[Shi Ziyuan] iP

shizy opened the pull request:
# Bobby
> ""You miss 100% of the shots you don't take" - Wayne Gretzky" - Michael Scott

Bobby is a **CLI program** that enables you to add and track various types of activities :sweat_smile:

Features:
- Task types (event, deadline, todo, etc.)
- Ability to mark/unmark tasks as complete
- Find functionality to help you locate tasks
- Listing ability to display all your tasks
- Persistence (saves to *local storage*)

Future add-ons:
- [ ] Nothing. It's ~~not yet~~ perfect
- [ ] Make it **sentient**

Steps for use:
1. Download the binary file (.jar) from [here](https://github.com/shizy/ip/releases) (download from the latest release)
2. Double-click on the .jar file
3. You can now use the program!

Here is a C Hello World program, just for fun:
``` C
#include <stdio.h>

int main(void) {
    printf("Hello World!");
    return 0;
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

shizy committed:
Renamed main class, added greetings

shizy committed:
Added echoing and the bye command

shizy committed:
Added list feature

shizy committed:
Refactored chatbot to utilise OOP

shizy committed:
Added Task class, added marking functionality

shizy committed:
Added Todo, Deadline and Event classes and support for these 3 tasks

shizy committed:
Added user input error handling with exceptions

shizy committed:
Added delete functionality, added more error handling

shizy committed:
Add support for reading in and creating a data file

shizy committed:
Add write to file functionality after any action

shizy committed:
Merge branch 'branch-Level-7'
Merge branch-level-7 into master branch

shizy committed:
Add proper date functionality for deadline tasks

shizy committed:
Add date functionality for event tasks

shizy committed:
Change LocalDate functionality to LocalDateTime for deadline and event tasks

shizy committed:
Merge branch 'branch-Level-8'
Merge branch-Level-8 into master

shizy committed:
Refactor code to separate logic into more distinct classes

shizy committed:
Merge branch 'refactor-more-oop'
Merge refactor-more-oop into master

shizy committed:
Refactor files to be in different packages

Clean up working directory, splitting different files with different purposes to different packages. This way, project becomes neater.

shizy committed:
Merge branch 'add-gradle-support'

shizy committed:
Add test cases using junit

Create test cases for tasks, Parser, TaskList and Ui components

shizy committed:
Add actual test cases with junit

Add actual test cases for the tasks, Parser, TaskList and Ui classes

shizy committed:
Change build.gradle config to build jar

shizy committed:
Add JavaDoc comments for the rest of the classes

Add JavaDoc comments for the Bobby, Parser, Storage, TaskList, Ui, and other task classes

shizy committed:
Minor change to comply with coding standards

Change made to boolean method name to comply with coding standards

shizy committed:
Modify JavaDoc comments to comply with coding standards

shizy committed:
Add find function and command to chatbot functionality

The find command allows users to input a query, and get the list of tasks that have names that encompass the query

shizy committed:
Merge branch 'branch-A-JavaDoc'
Merging branch-A-JavaDoc into master

shizy committed:
Fix merge conflict

shizy committed:
Fix merge conflict within Task.java

superb-sushi commented on src/main/java/chatbot/Bobby.java:
```diff
@@ -0,0 +1,54 @@
+    /** String constant representing the directory path of the file */
+    private static final String DIR_PATH = "./data/";
+    /** String constant representing the name of the file */
+    private static final String FILE_NAME = "tasks.txt";
```
I like the good use of capitalised variable names to represent constant variables!

superb-sushi commented on src/main/java/chatbot/Parser.java:
```diff
@@ -0,0 +1,162 @@
+package chatbot;
+
+import chatbot.exception.*;
```
Could this have been further specified instead of using a wildcard import?

superb-sushi commented on src/main/java/chatbot/Parser.java:
```diff
@@ -0,0 +1,162 @@
+    public static void processInput(String s, TaskList taskList, Storage storage, Ui ui) throws InputException {
+        String[] inputArr = s.split(" ", 2);
+        String command = inputArr[0];
+        switch (command) {
```
I like how you have made your "switch" statement indentations align with the given coding standards!

superb-sushi commented on src/main/java/chatbot/Storage.java:
```diff
@@ -0,0 +1,96 @@
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
```
Would it make things clearer to determine the relevant used dependencies if the different import statements where separated?
```suggestion
import java.io.IOException;

import java.util.ArrayList;
import java.util.Scanner;
```

superb-sushi commented on src/main/java/chatbot/TaskList.java:
```diff
@@ -0,0 +1,118 @@
+     */
+    public void find(String query) {
+        System.out.println("Here are the tasks matching your query:");
+        for (int i = 0; i < this.tasks.size(); i++) {
```
I like the use of whitespaces to wrap the semi-colons & operators!

superb-sushi reviewed (commented):
Overall, your code displayed good and consistent adherence to the given coding standards, barring minor mistakes! Very good job!

shizy committed:
Add checkstyle support, refactor code to be checkstyle compliant

Mainly refactored import orderings and certain whitespace issues

shizy committed:
Port over javafx tutorial GUI to chatbot

Chatbot and gui functionality is currently still separate

shizy committed:
Refactor chatbot functionality to comply with GUI

Changed many of the method return types to string, in order to connect the chatbot logic with the GUI logic

meatyturtle commented on src/main/java/chatbot/exception/DeadlineArgsException.java:
```diff
@@ -0,0 +1,10 @@
+package chatbot.exception;
```
i liked that you made specific exceptions for different cases, hope it helped with debugging!

meatyturtle commented on src/main/java/chatbot/logic/Parser.java:
```diff
@@ -0,0 +1,172 @@
+import chatbot.exception.MissingDeadlineByException;
+import chatbot.exception.MissingEventFromException;
+import chatbot.exception.MissingEventToException;
+import chatbot.task.Deadline;
```
you could leave a line in between chatbot.exception imports and chatbot.task imports for clarity

meatyturtle commented on src/main/java/chatbot/logic/Parser.java:
```diff
@@ -0,0 +1,172 @@
+        boolean isDone = Integer.parseInt(tokens[1]) == 1;
+        String title = tokens[2];
+        return switch (tokens[0]) {
+        case "T" -> new Todo(title, isDone);
```
i believe there should be 4(?) spaces of indentation for every line after `switch`

meatyturtle commented on src/main/java/chatbot/logic/Storage.java:
```diff
@@ -0,0 +1,96 @@
+        try {
+            Scanner fileScanner = this.getFile();
+            while (fileScanner.hasNextLine()) {
+                // I don't really like this nested try block
```
if it works, it works! 😝

meatyturtle commented on src/main/java/chatbot/logic/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+     * @throws InvalidIndexException Exception thrown if the index specified is invalid
+     */
+    public Task get(int idx) throws InvalidIndexException {
+        if (idx >= this.tasks.size() || idx < 0) {
```
good eye for error handling!

meatyturtle reviewed (commented):
looks good! the effort put in is evident 👍

shizy committed:
Add assertions to check assumption validity

This checks for assumptions and invariants that should hold at various points in the code

shizy committed:
Refactor code to ensure code quality

Reduced deep nesting, structured code logically with spacing, and reduced complicated expressions

shizy committed:
Add comments to improve code quality

shizy committed:
Change TaskList find and listTasks implementation from for loops to streams

While this technically overcomplicates the method implementations, since for loops are simpler and sufficient for the job, this was done to review the java stream concept

shizy committed:
Merge pull request #1 from shizy/branch-A-Assertions

Add assertions to check assumption validity

shizy committed:
Merge pull request #2 from shizy/branch-A-CodeQuality

Improve Code quality

shizy committed:
Resolve merge conflicts

shizy committed:
Merge pull request #3 from shizy/branch-A-Streams

Change methods to implement Java streams

shizy committed:
Add CI using GitHub Actions

shizy committed:
Remove unused imports

Removed unused imports from the Deadline and Event classes to fulfill checkstyle

shizy committed:
Add sort functionality to chatbot

Users can sort the task list with the sort command, with either the asc argument or the desc argument to sort in ascending and descending orders respectively

shizy committed:
Refactor Parser to utilise Command subclasses

The Command class and its subclasses hide the command execution logic from the Parser class. This helps to declutter the processInput() function in the Parser class

shizy committed:
Fix some bugs related to display of task strings

shizy committed:
Add Javadoc comments and fixed Javadoc formatting

shizy committed:
Add screenshot and user guide to README

README is meant for the users to understand the use cases and commands present in the program

shizy committed:
Replace profile images with different pictures

shizy committed:
Replace UI screenshot for User Guide

Updated UI screenshot to be consistent with the current version of the program
